File: frontend/src/types/agent.ts
// DevDocsAI V2 — Agent TypeScript Types

export type AgentMode = 'auto' | 'v2' | 'v1';

export type QueryType =
  | 'simple'
  | 'general_code'
  | 'architecture'
  | 'documentation'
  | 'code_review'
  | 'complex_multi_hop'
  | 'unknown'
  | 'error';

export interface AgentSourceReference {
  file: string;
  function?: string;
  line_start?: number;
  line_end?: number;
  content_preview?: string;
  score: number;
}

export interface AgentStep {
  agent_name: string;
  action: string;
  tools_invoked: string[];
  confidence: number;
  duration_ms: number;
  output_summary: string;
}

export interface ReasoningTrace {
  steps: AgentStep[];
  total_duration_ms: number;
  total_tokens_used: number;
  agents_invoked: string[];
  reflection_cycles: number;
  final_confidence: number;
}

export interface AgentChatResponse {
  answer: string;
  session_id: string;
  message_id: string;
  sources: AgentSourceReference[];
  reasoning_trace: ReasoningTrace;
  confidence: number;
  agents_invoked: string[];
  query_type: QueryType;
  model: string;
  mode: AgentMode;
  error?: string;
}

export interface AgentStatusEvent {
  type: 'agent_start' | 'agent_complete' | 'token' | 'done' | 'error';
  agent_name?: string;
  content?: string;
  confidence?: number;
  duration_ms?: number;
  step_index?: number;
  total_steps?: number;
  sources?: Record<string, unknown>[];
  reasoning_trace?: Record<string, unknown>;
}

export interface AgentModeInfo {
  id: AgentMode;
  name: string;
  description: string;
  default: boolean;
}

export interface AgentModesResponse {
  modes: AgentModeInfo[];
  current_mode: string;
  confidence_threshold: number;
  max_reflection_cycles: number;
}

// Agent display metadata for UI
export const AGENT_DISPLAY: Record<string, { label: string; icon: string; color: string }> = {
  orchestrator:   { label: 'Orchestrator',   icon: '🎯', color: '#6366F1' },
  planning:       { label: 'Planning',       icon: '📋', color: '#8B5CF6' },
  repository:     { label: 'Repository',     icon: '📦', color: '#06B6D4' },
  retrieval:      { label: 'Retrieval',      icon: '🔍', color: '#10B981' },
  architecture:   { label: 'Architecture',   icon: '🏛️', color: '#F59E0B' },
  documentation:  { label: 'Documentation',  icon: '📝', color: '#EC4899' },
  code_analysis:  { label: 'Code Analysis',  icon: '🔬', color: '#EF4444' },
  reflection:     { label: 'Reflection',     icon: '🪞', color: '#14B8A6' },
  aggregator:     { label: 'Aggregator',     icon: '⚡', color: '#F97316' },
  v1_fast_path:   { label: 'V1 Fast Path',  icon: '⚡', color: '#64748B' },
  v1_fallback:    { label: 'V1 Fallback',   icon: '🔄', color: '#64748B' },
};

File: backend/app/schemas/agent.py
"""
DevDocsAI V2 — Agent API Schemas
===================================
Request/response models for the V2 agent chat API.
These are separate from the V1 chat schemas to avoid breaking changes.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ── Request Schemas ──────────────────────────────────────────────────────────

class AgentChatRequest(BaseModel):
    """Request body for agent-powered chat queries."""

    repo_id: str = Field(..., description="Repository identifier")
    session_id: str = Field(..., description="Chat session identifier")
    question: str = Field(..., description="User's question about the repository")
    mode: str = Field(
        default="auto",
        description='Agent mode: "auto" (select based on complexity), "v2" (force multi-agent), "v1" (force V1 fast path)',
    )


# ── Response Schemas ─────────────────────────────────────────────────────────

class AgentStepOut(BaseModel):
    """One step in the reasoning trace displayed to the user."""

    agent_name: str
    action: str
    tools_invoked: List[str] = []
    confidence: float = 0.5
    duration_ms: int = 0
    output_summary: str = ""


class ReasoningTraceOut(BaseModel):
    """Full reasoning trace returned with every V2 response."""

    steps: List[AgentStepOut] = []
    total_duration_ms: int = 0
    total_tokens_used: int = 0
    agents_invoked: List[str] = []
    reflection_cycles: int = 0
    final_confidence: float = 0.5


class AgentSourceReference(BaseModel):
    """Source file reference from agent retrieval."""

    file: str
    function: Optional[str] = None
    line_start: Optional[int] = None
    line_end: Optional[int] = None
    content_preview: Optional[str] = None
    score: float = 0.0


class AgentChatResponse(BaseModel):
    """Response from the agent-powered chat pipeline."""

    answer: str = Field(..., description="AI-generated answer")
    session_id: str
    message_id: str
    sources: List[AgentSourceReference] = Field(
        default_factory=list, description="Source file references"
    )
    reasoning_trace: ReasoningTraceOut = Field(
        default_factory=ReasoningTraceOut,
        description="Full agent reasoning trace",
    )
    confidence: float = Field(
        default=0.5, ge=0.0, le=1.0, description="Final confidence score"
    )
    agents_invoked: List[str] = Field(
        default_factory=list, description="Names of agents that were invoked"
    )
    query_type: str = Field(
        default="unknown", description="Classified query type"
    )
    model: str = Field(default="", description="LLM model used")
    mode: str = Field(default="auto", description="Agent mode that was used")
    error: Optional[str] = Field(
        default=None, description="Error message if pipeline partially failed"
    )


# ── SSE Event Schemas ────────────────────────────────────────────────────────

class AgentStatusEvent(BaseModel):
    """Server-Sent Event for real-time agent status updates."""

    type: str = Field(
        ...,
        description=(
            'Event type: "agent_start", "agent_complete", "token", "done", "error"'
        ),
    )
    agent_name: Optional[str] = None
    content: Optional[str] = None
    confidence: Optional[float] = None
    duration_ms: Optional[int] = None
    step_index: Optional[int] = None
    total_steps: Optional[int] = None
    # Full response data (only in "done" events)
    sources: Optional[List[Dict[str, Any]]] = None
    reasoning_trace: Optional[Dict[str, Any]] = None
